docs/features/source-generators.cookbook.md

```diff
@@ -0,0 +1,250 @@
+# Source Generators Cookbook
+
+## Summary
+
+This document aims to be a guide to help the creation of source generators by providing a series of reusable patterns.
+It also aims to set out what types of generators are possible under the current design and what is explictly out of scope.
+
+> **Note**: The design for the source generator proposal is still under review. This document uses only one possible syntax, and
+> it is expected to change without notice as the feature evolves.
+
+## Proposal
+
+It is worth breifly setting out the design goals of generators here. For a more complete design see the [full design document](source-generators.md)
+
+- Generators produce one or more strings that represent C# source code to be added to the compilation.
+- Explicitly _additive_ only. Generators can add new code to a compilation but may **not** modify existing user code.
+- May access _additional files_, that is, non-C# source texts.
+- Run _un-ordered_, each generator will see the same input compilation, with no access to files created by other source generators.
+- A user specifies the generators to run via list of assemblies, much like analyzers.
+
+## Out of scope designs
+
+We will briefly look at the non-solvable problems as examples of the kind of problems source generators are *not* designed to solve:
+
+### Language features
+
+Source generators are not designed to replace new language features: for instance one could imagine [records](records.md) being implemented as a source generator
+that converts the specified syntax to a compilable C# representation.
+
+We explicitly consider this to be an anti-pattern; the language will continue to evolve and add new features, and we don't expect source generators to be
+a way to enable this. Doing so would create new 'dialects' of C# that are incompatible with the compiler without generators. Further, because generators,
+by design, cannot interact with each other, language features implemented in this way would quickly become incompatible with other additions to the language.
+
+### Code-not-source generators
+
+There are many post-processing tasks that users perform on their assemblies today, which here we define as 'code-not-source' generation. These include, but
+are not limited to:
+
+- Optimization
+- Logging injection
+- IL Weaving
+- Call site re-writing
+
+While these techniques have many valuable use cases, they do not fit into the idea of *source* generation. They are, by defintion, code altering operations
+which is explicitly ruled out by the source generation proposal.
+
+There are already well supported tools and techniques for acheiving these kinds of operations, and the source generators proposal is not aimed at replacing them.
+
+## Designs
+
+This section is broken down by user scenarios, with general solutions listed first, and more specific examples later on.
+
+### Generated class
```

Comment:
This makes sense. Have we thought about whether we should recommend any conventions? For example, should the generator honor the default namespace? Should the generator let the user control the namespace/type name? If so, how would we want them to express that configuration?

Additional files are powerful and so are attributes, but I'd prefer if we don't let everyone come up with a way to do this.

Reply:
Yep we definitely want to do this. I'm not sure we have the guidance yet, and I'm expecting we'll figure a lot of it out as we work with our first party customers. I'll make a note in this doc that we need to come back and add this.